cmd/posix.go

```diff
@@ -1243,41 +1217,57 @@ func (s *posix) CreateFile(volume, path string, fileSize int64, r io.Reader) (er
 		return remainingWritten, nil
 	}
 
-	var written int
+	bufp := s.pool.Get().(*[]byte)
+	defer s.pool.Put(bufp)
+
+	buf := *bufp
+
+	var written int64
 	for {
-		var n int
-		n, err = io.ReadFull(r, buf)
-		switch err {
-		case nil:
-			n, err = w.Write(buf)
-			if err != nil {
-				return err
+		nr, er := r.Read(buf)
+		if nr == len(buf) {
+			nw, ew := w.Write(buf)
+			if nw > 0 {
+				written += int64(nw)
 			}
-			written += n
-		case io.ErrUnexpectedEOF:
-			n, err = writeRemaining(w, buf[:n])
-			if err != nil {
-				return err
+			if ew != nil {
+				err = ew
+				break
 			}
-			written += n
-			fallthrough
-		case io.EOF:
-			if fileSize != -1 {
-				if written < int(fileSize) {
-					return errLessData
-				}
-				if written > int(fileSize) {
-					return errMoreData
-				}
+		} else if nr > 0 {
+			nw, ew := writeRemaining(w, buf[:nr])
+			if nw > 0 {
+				written += int64(nw)
+			}
+			if ew != nil {
+				err = ew
+				break
+			}
+			if nr != nw {
+				err = io.ErrShortWrite
+				break
 			}
-			return nil
-		default:
-			return err
 		}
+		if er != nil {
+			if er != io.EOF {
+				err = er
+			}
+			break
+		}
+	}
+
```

Comment:
shouldn't we prioritize errors received from the underlying reader and writer before returning error based on amount of data written to the file? 
```suggestion
if err != nil {
  return err
}
```